networking-calico/networking_calico/plugins/ml2/drivers/calico/mech_calico.py

```diff
@@ -790,8 +790,10 @@ def update_port_postcommit(self, context):
             endpoint_should_already_exist = port_bound(original)
 
             # Check for migration so that we can reliably delete the
-            # WorkloadEndpoint on the old host.
-            if original['binding:host_id'] != port['binding:host_id']:
+            # WorkloadEndpoint on the old host. That happens when the host
+            # formerly had a binding, and it doesn't match the current one.
+            if original['binding:host_id'] and \
+                    original['binding:host_id'] != port['binding:host_id']:
```

Comment:
Just thinking through this...

The change only makes a difference if `original['binding:host_id']` is Python-false.

It looks very plausible, but we use a different condition in all of the places where we're deciding to call `self.endpoint_syncer.write_endpoint`, namely `port_bound(port)` or `port['binding:vif_type'] != 'unbound'`.  Hence I think the proposed change is only correct if Neutron updates `port['binding:host_id']` and `port['binding:vif_type']` in perfect sync.  It feels like it would be safer to say `if endpoint_should_already_exist and` instead of `if original['binding:host_id'] and` - that should be correct in all cases, including when there is a perfect sync with `binding:host_id`.  WDYT?

Reply:
I agree with you as far as "And we try to delete something which does not exist."

My point, though, is that we only ever **create** an endpoint (in etcd) if `port_bound(port)`, so it follows that we only need to **delete** an endpoint if `port_bound(original)` is true for the previous port state `original`.  I think that should be 100% reliable so long as we don't miss any transitions between `port_bound(port)` and `!port_bound(port)`, and I've re-reviewed the code to check that it won't miss any such transitions.